[docs] Add a Linking section

amandeepmittal opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Fixes ENG-12865 ENG-11847

# How

<!--
How did you build this feature or fix this bug and why?
-->

As per the issue, organize the linking and deep linking docs by creating a section called "Linking".

This section contains the following pages:

**Overview**
Preview Link: /linking/overview/

- Provide a description of what linking is, explain what a linking URL looks like
- Explain different types of linking strategies (Into other apps, into your app, Android App Links, iOS Universal links)
- Explain the benefits of using Expo Router

**Linking into other apps** 
Preview link: /linking/into-other-apps/

The page that describes the following:
- Introduction to answer how linking to other apps is achieved
- Provide info on opening common URL schemes using `expo-linking` API and `Link` component when using Expo Router
- What are common URL schemes
  - Specify Android intents by creating a config plugin. This is applicable for Android 11 and above: https://reactnative.dev/docs/linking?syntax=android#canopenurl
- Custom URL schemes
- Create URLs
- Removed Expo published info for Expo Go since publishing Expo Go apps is deprecated and removed
- Organize Expo Go info under a Collapsible
- Using in-app browsers with `expo-web-browser`. Organize example under Collapsible
- Implementing additional link functionality on the web either using `Link` from Expo Router or `A` from `@expo/html-elements`

**Linking into your app**
Preview link: /linking/into-your-app/

The page that describes the following:
- Add a `scheme` in app config. Add a note that Expo Router enables this automatically and user can skip adding the `scheme` because `android.package` and `ios.bundleIdentifier` can be used
- Test a link in development using `npx uri-scheme` tool. Organize Expo Go info under a Collapsible and remove published Expo Go related info since that is deprecated and removed
- Handle links using `Linking.useURL()` hook with an example
- Parse URLs using `Linking.parse()` method
- When not to use deep links

**Android App Links**
Preview link: /linking/android-app-links/

This page describes the following:
 
- Step to add `intentFilters` to app config and enable `autoVerify` to true
- Set up two-way association for website verification and native app verification (Create `assetlinks.json` file, host the file, and native app verification process description)
- Debug Android App Links locally using Expo CLI and `adb` shell command
- Troubleshooting tips section
- Removed workaround for `expo-dev-client` version `1.2x` and below to modify AndroidManifest.xml file and remove `exp+slug` instructions after testing with the latest `expo-dev-client` version and Android 14.

**iOS Universal Links**
Preview link: /linking/ios-universal-links/

This page describes the following:

- Set up two-way association for website verification and native app verification (Create AASA file, host the file, Native app configuration and verification steps and description)
- Apple Smart Banner
- Debug iOS Universal Links locally using Expo CLI
- Troubleshooting tips section
- Removed the note about implementing deferred links using `react-native-branch` (cc: @brentvatne, not sure if this holds any value in this guide anymore)

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

amandeepmittal committed:
Add Linking section and Linking into other apps page

amandeepmittal committed:
Update description and Custom URL schemes section in "Linking into other apps"

amandeepmittal committed:
Fix section title and add link to API doc

amandeepmittal committed:
Add "Linking into your app"

amandeepmittal committed:
"Linking into your app": Update section titles and improve code snippets

amandeepmittal committed:
Update into-your-app.mdx

amandeepmittal committed:
Add "Create URLs" section in "Linking into other apps"

amandeepmittal committed:
Update into-other-apps.mdx

amandeepmittal committed:
Update into-other-apps.mdx

amandeepmittal committed:
Update into-your-app.mdx

amandeepmittal committed:
Linking into your app: Improve section headings, description, add When not use deep links section

amandeepmittal committed:
Add "Android App Links"

amandeepmittal committed:
Android App Links: update description & intro, fix link

amandeepmittal committed:
Update android-app-links.mdx

amandeepmittal committed:
Update android-app-links.mdx

amandeepmittal committed:
Add iOS Universal Links

amandeepmittal committed:
Create an overview page, remove previous Linking & Deep Linking pages

amandeepmittal committed:
Add redirects for deleted URLs

amandeepmittal committed:
Android App Links: Fix description

amandeepmittal committed:
Overview: Add description

amandeepmittal committed:
Complete: Overview, Into other apps, Into your app

amandeepmittal committed:
Overview: Fix internal links

amandeepmittal committed:
Update all pages

amandeepmittal commented on docs/pages/linking/into-your-app.mdx:
```diff
@@ -0,0 +1,129 @@
+
+## When to not use deep links
+
+If a user does not has your app installed and follows a link to your app's custom scheme, the device's operating system will indicate that the page couldn't be opened without providing any further information. This results in a poor user experience, and unfortunately, there's no way to work around this in a web browser.
```
I'm not sure (after working on iOS universal links and Android App Links), the last statement is true anymore, since Apple Smart Banner handles this for iOS. I think I might not have complete context.

@brentvatne: is the last statement still relevant?

keith-kurak replied:
We could walk around this a bit by being less prescriptive ... instead of "when not to use deep links", this could be "limitations" and something as simple as, "If the user doesn't have your app installed, deep links don't link to anything." and optionally "Attribution services like Branch offer solutions for conditionally linking to your app or a web page" (at least as I recall) and/ or "one solution is to use universal links, which will route to your web page when the app isn't installed [learn more]"

amandeepmittal committed:
Fix CI error

amandeepmittal committed:
Overview: Add a section about Expo Router

amandeepmittal committed:
Fix typo

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,108 @@
+
+### Into your app
+
+Linking to your app from other apps or a web browser is achieved using a URL based on the scheme for your app. These URLs are called **deep links**.
```
The document introduces "links" and "deep links" in the title. This is the first time afterwards where it references "deep links". If we're presenting these as two concepts (which I think is useful on an overview page like this), let's differentiate them more strongly. I think this definition is a good one to anchor this to: https://en.wikipedia.org/wiki/Deep_linking. Deep links aren't specific to apps; they're just links to specific url-based content inside of a website or app. 

Thinking back to apps, any app with a registered scheme can be linked to (e.g., `myapp://whatever`). Only apps that listen for deep links and perform routing based on them really implement "deep linking", and it just happens that Expo Router makes this happen automatically. (seems like long ago, but a lot of web apps used to not support deep linking... actually, many still don't).

keith-kurak commented on docs/pages/linking/overview.mdx:
````diff
@@ -0,0 +1,108 @@
+
+For example, by clicking an advertisement of your product, your app will open on the user's device and they can view that product's details. This product's link that the user clicked may look like:
+
+```html
````
See comment below; this is actually a deep link, I believe.

amandeepmittal replied:
I've moved this example to Deep links section

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,108 @@
+  Icon={BookOpen02Icon}
+/>
+
+### Android App Links
```
Can we group app links and universal links under a single heading, and then break out in much more brief into each platform? The story for each is basically the same: "This is how you route an http(s) link to a particular domain to your app if it is installed", just they're named differently per platform. I liken these to "App Shortcuts / Quick Actions": same concept that we'd probably call by the same name if Google and Apple could settle on one, and even then, we probably just pick the name from one platform and use it colloquially to refer to both.

amandeepmittal replied:
We used to use Universal links before this PR made a division. Do you want me to still use "Universal links"?

amandeepmittal replied:
Here's the update based on your initial feedback:

![CleanShot 2024-09-23 at 19 27 20](https://github.com/user-attachments/assets/14b9056f-b301-4f32-a891-91ec043221ab)

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,108 @@
+
+## Use Expo Router to handle deep linking
+
+To implement any of the above Linking strategy, **we recommend using** [Expo Router](/router/introduction/) since deep linking is automatically enabled for all of your app's screens.
```
```suggestion
To implement any of the above Linking strategies, **we recommend using** [Expo Router](/router/introduction/) since deep linking is automatically enabled for all of your app's screens.
```

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,108 @@
+
+There are different linking strategies you handle in your Expo app:
+
+- Linking to other apps from your app (outgoing links)
```
I might organize these differently to reinforce the differences between linking, deep linking, and the similarities between app links and universal links:
```
- Linking to other apps from your app (outgoing links)
- Linking to your app from other apps or a web browser (incoming links)
- Linking to specific content within your app (deep links)
- Redirecting links to your web domain to your app if it is installed (Android App Links and iOS Universal Links)
```

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,108 @@
+
+**Benefits:**
+
+- When implementing deep links, you can skip [adding `scheme` to your app's config](/linking/into-your-app/#add-a-custom-scheme-in-app-config).
```
I'm fuzzy about how this is a benefit of Expo Router. All Expo apps register the bundle ID as a default scheme, I believe. Router helps define the url after the scheme, but doesn't affect the scheme itself, AFAIK.

amandeepmittal replied:
I'll remove it from here but on testing this, I found if you are using Expo Router, there's no requirement to use `scheme` and instead, an app can use ios bundler identifier or android package name to create deep links.

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,108 @@
+
+- When implementing deep links, you can skip [adding `scheme` to your app's config](/linking/into-your-app/#add-a-custom-scheme-in-app-config).
+- `Link` component from Expo Router can be used to [handle URL schemes to other apps](/linking/into-other-apps/#expo-router)
+- Android App Links and iOS Universal Links require configuring two-way association and runtime routing for the link in your app. This configuration is done in JavaScript and must be applied to every route to ensure coordination between web and native platforms. Since Expo Router enables deep links automatically, you don't have to configure runtime routing separately.
```
Also fuzzy on this being a router-specific benefit ... Router requires the same universal link configuration in order for universal links to work. I guess one indirect benefit is that if your bespoke non-router link handling matched on the scheme as well as the deep link, that would break with a universal link (or any other scheme, for that matter).

amandeepmittal replied:
This piece of content was already there in both sections (App and Universal links) as a note and was presented as a benefit of Expo Router. Instead of repeating ourselves in both documents, I decided to add this here as one place to gather all the info.

Based on your feedback, should I re-word this to say "hey, this is an indirect benefit of using Expo Route". On my research, I was also comparing this React navigation where you have to define a prefix: https://reactnavigation.org/docs/deep-linking/. iiuc, this is handled by Expo Router automatically.

keith-kurak commented on docs/pages/linking/into-your-app.mdx:
```diff
@@ -0,0 +1,127 @@
+
+## When to not use deep links
+
+If a user does not has your app installed and follows a link to your app's custom scheme, the device's operating system will indicate that the page couldn't be opened without providing any further information. This results in a poor user experience, and unfortunately, there's no way to work around this in a web browser.
```
```suggestion
If a user does not have your app installed and follows a link to your app's custom scheme, the device's operating system will indicate that the page couldn't be opened without providing any further information. This results in a poor user experience, and unfortunately, there's no way to work around this in a web browser.
```

keith-kurak commented on docs/pages/linking/into-your-app.mdx:
````diff
@@ -0,0 +1,127 @@
+}
+```
+
+## When to not use deep links
````
Kind of goes for the entire page, but if we update the overview to clarify a link vs deep link, might be good to sync this language, because a lot of this is talking about links that may or may not be deep.

keith-kurak commented on docs/pages/linking/into-your-app.mdx:
```diff
@@ -0,0 +1,127 @@
+
+## Handle URLs
+
+You can observe links that launch your app using the [`Linking.useURL()`](/versions/latest/sdk/linking/#useurl) hook from [`expo-linking`](/versions/latest/sdk/linking/).
```
might be worth a call-out to the effect of "you can ignore all this with Expo Router"

amandeepmittal committed:
Apply feedback from Keith

keith-kurak commented on docs/pages/linking/into-other-apps.mdx:
```diff
@@ -0,0 +1,261 @@
+
+## In-app browsers
+
+The `expo-linking` API allows you to open a URL using the operating system's default or preferred web browser app. You can use the [`expo-web-browser`](/versions/latest/sdk/webbrowser) library to open URLs in an in-app browser. For example, an in-app browser is useful for secure [authentication](/guides/authentication).
```
```suggestion
The `expo-linking` API allows you to open a URL using the operating system's default web browser app. You can use the [`expo-web-browser`](/versions/latest/sdk/webbrowser) library to open URLs in an in-app browser. For example, an in-app browser is useful for secure [authentication](/guides/authentication).
```

keith-kurak commented on docs/pages/linking/into-your-app.mdx:
```diff
@@ -0,0 +1,119 @@
+---
+title: Linking into your app
+sidebar_title: Into your app
+description: Learn how to open and handle a URL from other apps based on the scheme for your by creating a deep link.
```
```suggestion
description: Learn how to handle an incoming URL in your app
```
Realizing maybe there was some search optimization happening here, so feel free to add back words

keith-kurak commented on docs/pages/linking/into-your-app.mdx:
```diff
@@ -0,0 +1,119 @@
+
+## Limitations
+
+If a user does not have your app installed, deep links don't link to anything. Attribution services like [Branch](https://www.branch.io/deep-linking/) offer solutions for condistionally linking to your app or web page.
```
```suggestion
If a user does not have your app installed, deep links to your app will not work. Attribution services like [Branch](https://www.branch.io/deep-linking/) offer solutions for condistionally linking to your app or web page.
```

keith-kurak commented on docs/pages/linking/into-your-app.mdx:
```diff
@@ -0,0 +1,119 @@
+
+If a user does not have your app installed, deep links don't link to anything. Attribution services like [Branch](https://www.branch.io/deep-linking/) offer solutions for condistionally linking to your app or web page.
+
+Universal links is another solution you can use to handle such cases, in which the route to your web page determins when the app is installed or not. For more details, see [Universal Links](/linking/ios-universal-links/).
```
```suggestion
Universal links is another solution you can use to handle such cases, in which the route to your web page determines whether or not the app is installed. For more details, see [Universal Links](/linking/ios-universal-links/).
```

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,109 @@
+  Icon={BookOpen02Icon}
+/>
+
+## App shortcuts
```
```suggestion
## Universal linking

Both Android and iOS implement their own systems for routing web URL's to an app if the app is installed. On Android, this system is called App Links, and on iOS it is called Universal Links.
```
Maybe something like this? I do tend to see this referred to as "universal links" or "universal linking" colloquially.

amandeepmittal replied:
Yeah, we used to call the group "universal links" before. Sounds good to me!

amandeepmittal commented on docs/pages/linking/android-app-links.mdx:
```diff
@@ -0,0 +1,177 @@
+
+### Native app verification
+
+Install the app on an Android device to trigger the [Android app verification](https://developer.android.com/training/app-links/verify-android-applinks#web-assoc) process. This process takes at least 20 seconds.
```
Agreed!

amandeepmittal commented on docs/pages/linking/into-other-apps.mdx:
```diff
@@ -0,0 +1,261 @@
+
+## Common URL schemes
+
+There are built-in URL schemes that provide access to core functionality on all platforms. Here is a list of commonly used schemes:
```
It is.

amandeepmittal commented on docs/pages/linking/into-other-apps.mdx:
```diff
@@ -0,0 +1,261 @@
+
+> If you know the custom scheme for the app you want to open from your app, you can link to it using any of the methods: [Using `expo-linking` API](#using-expo-linking-api) or [Using `Link` from Expo Router](#using-expo-routers-link-component).
+
+Some services provide documentation on how to handle their app's custom URL schemes. For example, [Uber's deep linking documentation](https://developer.uber.com/docs/riders/ride-requests/tutorials/deep-links/introduction#standard-deep-links) describes how to link directly to a specific pickup location and destination:
```
Makes sense

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Keith Kurak <keith.kurak@gmail.com>

amandeepmittal commented on docs/pages/linking/into-your-app.mdx:
```diff
@@ -0,0 +1,119 @@
+---
+title: Linking into your app
+sidebar_title: Into your app
+description: Learn how to handle an incoming URL in your app
```
```suggestion
description: Learn how to handle an incoming URL in your app by creating a deep link.
```

amandeepmittal committed:
Apply suggestions from code review

amandeepmittal committed:
Update overview.mdx

amandeepmittal committed:
[docs] Clean up old internal links of Linking and deep linking guides (#31283)

<!--
Please describe the motivation for this PR, and link to relevant GitHub
issues, forums posts, or feature requests.
-->

Follow up #30958

<!--
How did you build this feature or fix this bug and why?
-->

- Clean up and update internal URLs for previously existing Linking and
Deep Linking guides.
- Delete the previously existing `linking.mdx` and `deep-linking.mdx`
files
- Fix redirects introduced in #30958

<!--
Please describe how you tested this change and how a reviewer could
reproduce your test, especially if this PR does not include automated
tests! If possible, please also provide terminal output and/or
screenshots demonstrating your test/reproduction.
-->

N/A

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).

keith-kurak commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,111 @@
+
+iOS Universal Links are different from [standard deep links](#into-your-app) as they use regular HTTP and HTTPS schemes and are exclusive to iOS devices.
+
+This link type allows your app to always open when a user clicks the link instead of choosing between the browser or another handler from a dialog displayed on the device. If the user doesn't have your app installed, the link takes them to your app's associated website. You can further configure the website by displaying a banner to the user to open your app using [Apple Smart Banner](/linking/io ...
```
```suggestion
This link type allows your app to open when a user clicks an HTTP(S) link pointing to your web domain. If the user doesn't have your app installed, the link takes them to your app's associated website. You can further configure the website by displaying a banner for the user to open your app using [Apple Smart Banner](/linking/ios-universal-links/#apple-smart-banner).
```
I don't use iOS as much, but I don't recall iOS prompting for how to handle a link. Usually one app clobbers another if they're sitting on the same scheme, and I thought that an https link would just go to the browser if there's no universal linking app registered

amandeepmittal replied:
Yes you are right. On IOS, they will be directed to the website. I think I had the wrong context here.

keith-kurak reviewed (approved):
Looks good! Just noted a few small grammar bits and maybe one more question about universal links

amandeepmittal commented on docs/pages/linking/overview.mdx:
```diff
@@ -0,0 +1,111 @@
+
+Linking to other apps from your app is achieved using a URL based on the target app's URL scheme. This **URL scheme** allows you to reference resources within that native app.
+
+Your app can use a [common URL scheme](/linking/into-other-apps/#common-url-schemes) for default apps or implement any URL scheme, including `https` and `http` (commonly used by web browsers like Chrome, Safari, and so on), and use JavaScript to handle the URL that launches the corresponding native app.
```
You are right, yes invoking the URL is the action that the user can take here by following the link below.

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Keith Kurak <keith.kurak@gmail.com>

amandeepmittal committed:
Delete react-native

amandeepmittal committed:
Revert "Delete react-native"

This reverts commit a3e44a31c63d073fecc289280216557a0b870ff8.

amandeepmittal commented:
@keith-kurak, thanks for the review and the approval. I am going to move all of our changes to a new branch and merge it, and then close this branch because I'm not sure why react-native submodules change --after doing a rebase-- are introduced in this PR and not sure if they cause a problem with the `main` branch or not.
